subprojects/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/processor/AbstractInstrumentationProcessor.java

```diff
@@ -73,11 +79,35 @@ private Set<Class<? extends Annotation>> getSupportedAnnotations() {
 
     @Override
     public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
-        Set<? extends Element> annotatedTypes = getSupportedAnnotations().stream().flatMap(annotation -> roundEnv.getElementsAnnotatedWith(annotation).stream()).collect(Collectors.toSet());
+        Set<? extends Element> annotatedTypes = getSupportedAnnotations().stream()
+            .flatMap(annotation -> roundEnv.getElementsAnnotatedWith(annotation).stream())
+            .collect(Collectors.toSet());
+        annotatedTypes = findActualTypesToVisit(annotatedTypes);
         collectAndProcessRequests(annotatedTypes);
         return true;
     }
 
+    private Set<? extends Element> findActualTypesToVisit(Set<? extends Element> annotatedTypes) {
+        return annotatedTypes.stream()
+            .flatMap(it -> findActualTypesToVisit(it).stream())
+            .collect(Collectors.toSet());
+    }
+
+    private Set<Element> findActualTypesToVisit(Element typeElement) {
+        Optional<? extends AnnotationMirror> annotationMirror = AnnotationUtils.findAnnotationMirror(typeElement, VisitForInstrumentation.class);
+        if (!annotationMirror.isPresent()) {
+            return Collections.singleton(typeElement);
+        }
+
+        @SuppressWarnings("unchecked")
+        List<AnnotationValue> values = (List<AnnotationValue>) AnnotationUtils.findAnnotationValue(annotationMirror.get(), "value")
+            .orElseThrow(() -> new IllegalStateException("missing annotation value"))
+            .getValue();
+        return values.stream()
+            .map(v -> processingEnv.getTypeUtils().asElement((TypeMirror) v.getValue()))
+            .collect(Collectors.toSet());
```

Comment:
Note that `Collectors.toSet()` gives non-deterministic ordering of the results. Please consider collecting the results to a list or an ordered set using `toCollection()` instead.

Reply:
(see https://github.com/gradle/gradle/pull/24632#discussion_r1171274638)